asset-transfer-events/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright IBM Corp. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as grpc from '@grpc/grpc-js';
+import { ChaincodeEvent, CloseableAsyncIterable, connect, Contract, GatewayError, Network } from '@hyperledger/fabric-gateway';
+import { TextDecoder } from 'util';
+
+import { newGrpcConnection, newIdentity, newSigner } from './connect';
+
+const channelName = 'mychannel';
+const chaincodeName = 'events';
+
+// Path to crypto materials.
+
+const utf8Decoder = new TextDecoder();
+const now = Date.now();
+const assetId = `asset${now}`;
+
+
+async function main(): Promise<void> {
+    // The gRPC client connection should be shared by all Gateway connections to this endpoint.
+    const client = await newGrpcConnection();
+
+    const gateway = connect({
+        client,
+        identity: await newIdentity(),
+        signer: await newSigner(),
+        evaluateOptions: () => {
+            return { deadline: Date.now() + 5000 }; // 5 seconds
+        },
+        endorseOptions: () => {
+            return { deadline: Date.now() + 15000 }; // 15 seconds
+        },
+        submitOptions: () => {
+            return { deadline: Date.now() + 5000 }; // 5 seconds
+        },
+        commitStatusOptions: () => {
+            return { deadline: Date.now() + 60000 }; // 1 minute
+        },
+    });
+
+    let events: CloseableAsyncIterable<ChaincodeEvent> | undefined;
+
+    try {
+        // Get a network instance representing the channel where the smart contract is deployed.
+        const network = gateway.getNetwork(channelName);
+
+        // Get the smart contract from the network.
+        const contract = network.getContract(chaincodeName);
+
+        //Start Listening to events.
+        events = await startEventListening(network);
+
+        // Create a new asset on the ledger.
+        const firstBlockNumber = await createAsset(contract);
+
+        // Update an existing asset.
+        await updateAsset(contract)
+
+        // Transfer an existing asset.
+        await transferAsset(contract);
+
+        // Delete asset by assetID.
+        await deleteAssetByID(contract);
+
+        // Replay all the events received.
+        await Promise.race([timeout(10000),replayChaincodeEvents(network,firstBlockNumber)])
```

Comment:
I appreciate you're replicating the read timeout in the Go sample. That was more to demonstrate different patterns of Go channel usage for event processing, and was a per-event timeout rather than for the whole sequence of reads.

If you want to keep this timeout example (which is reasonable), you probably need to change the logical flow slightly so you have the `CloseableAsyncIterable<ChaincodeEvent>` (or at least the `close()` function from it) from this event reading session and can close it if the timeout fires.

Implementing completely cleanly is going to be a little more complex so you might want to take one of these alternative approaches:

1. Set a `deadline` CallOption for the call to get chaincode events, so gRPC will timeout the connection at the deadline for you.
2. Don't worry about the timeout.

It's good practice to have some kind of timeout so the client application won't ever hang indefinitely but, if it's looking complicated to achieve that, it might be better to keep things simple so the sample code is as readable as possible.